tools/cmt2annot.ml

```diff
@@ -153,33 +168,58 @@ let binary_part iter x =
   | Partial_signature_item x -> app iter.signature_item x
   | Partial_module_type x -> app iter.module_type x
 
-let gen_annot target_filename filename
-              {Cmt_format.cmt_loadpath; cmt_annots; cmt_use_summaries; _} =
+(* Save cmt information as faked annotations, attached to
+   Location.none, on top of the .annot file. Only when -save-cmt-info is
+   provided to ocaml_cmt.
+*)
+let record_cmt_info cmt =
+  let location_none = {
+    Location.none with Location.loc_ghost = false }
+  in
+  let location_file file = {
+    Location.none with
+      Location.loc_start = {
+        Location.none.Location.loc_start with
+          Lexing.pos_fname = file }}
+  in
+  let record_info name value =
+    let ident = Printf.sprintf ".%s" name in
+    Stypes.record (Stypes.An_ident (location_none, ident,
+                                    Annot.Idef (location_file value)))
+  in
+  let open Cmt_format in
+  (* record in reverse order to get them in correct order... *)
+  List.iter (fun dir -> record_info "include" dir) (List.rev cmt.cmt_loadpath);
+  record_info "chdir" cmt.cmt_builddir;
+  (match cmt.cmt_sourcefile with
+    None -> () | Some file -> record_info "source" file)
+
+let gen_annot ?(save_cmt_info=false) target_filename filename cmt =
   let open Cmt_format in
   Envaux.reset_cache ();
-  Config.load_path := cmt_loadpath;
+  Config.load_path := cmt.cmt_loadpath @ !Config.load_path;
   let target_filename =
     match target_filename with
     | None -> Some (filename ^ ".annot")
     | Some "-" -> None
     | Some _ -> target_filename
   in
-  let iterator = iterator ~scope:Location.none cmt_use_summaries in
-  match cmt_annots with
+  if save_cmt_info then record_cmt_info cmt;
+  let iterator = iterator ~scope:Location.none cmt.cmt_use_summaries in
+  match cmt.cmt_annots with
   | Implementation typedtree ->
-      ignore (iterator.structure iterator typedtree);
-      Stypes.dump target_filename
+    ignore (iterator.structure iterator typedtree);
+    Stypes.dump target_filename
   | Interface _ ->
-      Printf.eprintf "Cannot generate annotations for interface file\n%!";
-      exit 2
+    Printf.eprintf "Cannot generate annotations for interface file\n%!";
+    exit 2
   | Partial_implementation parts ->
-      Array.iter (binary_part iterator) parts;
-      Stypes.dump target_filename
-  | _ ->
-      Printf.fprintf stderr "File was generated with an error\n%!";
-      exit 2
-
-
+    Array.iter (binary_part iterator) parts;
+    Stypes.dump target_filename
+  | Packed _ ->
+    Printf.fprintf stderr "Packed files not yet supported\n%!";
+    Stypes.dump target_filename
+  | Partial_interface _ -> assert false
```

Comment:
I realise there are differences of opinion on how `match` blocks should be indented, but this would be considerably easier to review if the new match case followed the 4 spaces indentation of the existing clauses, rather than altering everything (as it happens, I think the existing indentation is better, as it more clearly differentiates the match clauses from the blocks of code).